hal/src/i2c.rs

```diff
@@ -0,0 +1,412 @@
+//! # Inter-Integrated Circuit (I2C) bus
+
+use core::convert::TryFrom;
+
+use crate::{
+    embedded_hal::blocking::i2c::{Read, Write, WriteRead},
+    pac::{rcc::ccipr::I2C3SEL_A, I2C1, I2C2, I2C3, RCC},
+    rcc::{sysclk_hz, pclk1_hz},
+};
+
+use embedded_time::{fixed_point::FixedPoint, rate::*};
+
+/// I2C error
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum Error {
+    /// Arbitration loss
+    Arbitration,
+    /// Bus error
+    Bus,
+    /// Bus busy
+    Busy,
+    /// Not Acknowledge received
+    Nack,
+    /* Overrun, // slave mode only
+     * Pec, // SMBUS mode only
+     * Timeout, // SMBUS mode only
+     * Alert, // SMBUS mode only */
+}
+
+/// I2C peripheral operating in master mode
+pub struct I2c1<PINS> {
+    i2c: I2C1,
+    pins: PINS,
+}
+
+pub struct I2c2<PINS> {
+    i2c: I2C2,
+    pins: PINS,
+}
+
+pub struct I2c3<PINS> {
+    i2c: I2C3,
+    pins: PINS,
+}
+
+macro_rules! busy_wait {
+    ($i2c:expr, $flag:ident, $variant:ident) => {
+        loop {
+            let isr = $i2c.isr.read();
+            let icr = &$i2c.icr;
+
+            if isr.arlo().is_lost() {
+                icr.write(|w| w.arlocf().clear());
+                return Err(Error::Arbitration);
+            } else if isr.berr().is_error() {
+                icr.write(|w| w.berrcf().clear());
+                return Err(Error::Bus);
+            } else if isr.nackf().is_nack() {
+                while $i2c.isr.read().stopf().is_no_stop() {}
+                icr.write(|w| w.nackcf().clear());
+                icr.write(|w| w.stopcf().clear());
+                return Err(Error::Nack);
+            } else if isr.$flag().$variant() {
+                break;
+            }
+        }
+    };
+}
+
+macro_rules! i2c {
+    ($($I2cX:ident: ($I2CX:ident, $i2cXen:ident, $i2cXrst:ident, $i2cXsel:ident,
+                     $I2cXSda:ident, $I2cXScl:ident, $i2cXsclAf:ident, $i2cXsdaAf:ident),)+) => {
+        $(
+            impl<SCL, SDA> $I2cX<(SCL, SDA)> {
+                /// Enables clock and resets the peripheral
+                fn enable_clock(rcc: &mut RCC) {
+                    rcc.apb1enr1.modify(|_, w| w.$i2cXen().enabled());
+                    rcc.apb1rstr1.modify(|_, w| w.$i2cXrst().reset());
+                    rcc.apb1rstr1.modify(|_, w| w.$i2cXrst().clear_bit());
```

Comment:
This should be in a `pulse_reset` function for consistency with the other peripheral drivers.